src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -378,124 +405,160 @@ senderBalance < (UInt256)transaction.GasLimit * transaction.MaxFeePerGas + value
                             unspentGas -= codeDepositGasCost;
                         }
                     }
 
                     foreach (Address toBeDestroyed in substate.DestroyList)
                     {
-                        if (_logger.IsTrace) _logger.Trace($"Destroying account {toBeDestroyed}");
+                        if (_logger.IsTrace)
+                            _logger.Trace($"Destroying account {toBeDestroyed}");
+
                         _worldState.ClearStorage(toBeDestroyed);
                         _worldState.DeleteAccount(toBeDestroyed);
-                        if (txTracer.IsTracingRefunds) txTracer.ReportRefund(RefundOf.Destroy(spec.IsEip3529Enabled));
+
+                        if (tracer.IsTracingRefunds)
+                            tracer.ReportRefund(RefundOf.Destroy(spec.IsEip3529Enabled));
                     }
 
                     statusCode = StatusCode.Success;
                 }
 
-                spentGas = Refund(gasLimit, unspentGas, substate, caller, effectiveGasPrice, noValidation, spec);
+                spentGas = Refund(tx.GasLimit, unspentGas, substate, tx.SenderAddress, env.TxExecutionContext.GasPrice, opts, spec);
             }
-            catch (Exception ex) when (
-                ex is EvmException || ex is OverflowException) // TODO: OverflowException? still needed? hope not
+            catch (Exception ex) when (ex is EvmException || ex is OverflowException) // TODO: OverflowException? still needed? hope not
             {
-                if (_logger.IsTrace) _logger.Trace($"EVM EXCEPTION: {ex.GetType().Name}");
+                if (_logger.IsTrace)
+                    _logger.Trace($"EVM EXCEPTION: {ex.GetType().Name}");
                 _worldState.Restore(snapshot);
             }
 
-            if (_logger.IsTrace) _logger.Trace("Gas spent: " + spentGas);
+            if (!opts.HasFlag(ExecutionOptions.NoValidation) && !tx.IsSystem())
+                blk.GasUsed += spentGas;
 
-            Address gasBeneficiary = block.GasBeneficiary;
-            bool gasBeneficiaryNotDestroyed = substate?.DestroyList.Contains(gasBeneficiary) != true;
-            if (statusCode == StatusCode.Failure || gasBeneficiaryNotDestroyed)
-            {
-                if (notSystemTransaction)
-                {
-                    UInt256 fees = (ulong)spentGas * premiumPerGas;
-                    if (_worldState.AccountExists(gasBeneficiary))
-                    {
-                        _worldState.AddToBalance(gasBeneficiary, fees, spec);
-                    }
-                    else
-                    {
-                        _worldState.CreateAccount(gasBeneficiary, fees);
-                    }
+            return true;
+        }
 
-                    UInt256 burntFees = !transaction.IsFree() ? (ulong)spentGas * block.BaseFeePerGas : 0;
+        protected virtual void Execute(Transaction tx, BlockHeader blk, ITxTracer tracer, ExecutionOptions opts)
```

Comment:
Minor: I would put this method as first to improve code readability, as it is calling all other methods